File: services/sales_service/schemas/sales_note.py
"""
Pydantic schemas for Sales Notes CRUD.
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class SalesNoteItemCreate(BaseModel):
    """Schema for a line item when creating a note."""
    product_id: Optional[int] = None
    sku: Optional[str] = Field(None, max_length=100)
    product_name: str = Field(..., min_length=1, max_length=200)
    unit_measure: str = Field("LITROS", max_length=20)
    quantity: float = Field(..., gt=0)
    unit_price: float = Field(..., ge=0)
    discount_pct: float = Field(0, ge=0, le=100)
    ieps_rate: float = Field(0, ge=0)
    iva_rate: float = Field(0.16, ge=0)


class SalesNoteCreate(BaseModel):
    """Schema for creating a sales note."""
    client_id: Optional[int] = None
    client_name: Optional[str] = Field(None, max_length=200)
    channel: str = Field("B2B", max_length=20)
    payment_method: str = Field("TRANSFERENCIA", max_length=30)
    include_taxes: bool = Field(False, description="When False, IEPS/IVA shown as empty (not invoiced)")
    notes: Optional[str] = Field(None, max_length=500)
    created_by: Optional[str] = Field(None, max_length=100)
    items: list[SalesNoteItemCreate] = Field(..., min_length=1)

    class Config:
        json_schema_extra = {
            "example": {
                "client_id": 1,
                "client_name": "Hop Experience",
                "channel": "B2B",
                "payment_method": "TRANSFERENCIA",
                "include_taxes": False,
                "items": [
                    {
                        "product_name": "Litro Carnotaurus (Amber Lager)",
                        "unit_measure": "LITROS",
                        "quantity": 18,
                        "unit_price": 77.44,
                    },
                    {
                        "product_name": "Litro Coahuilaceratops (Imperial Indian Pale Ale)",
                        "unit_measure": "LITROS",
                        "quantity": 18,
                        "unit_price": 109.44,
                    },
                    {
                        "product_name": "Barril Desechable 20L",
                        "unit_measure": "UNITS",
                        "quantity": 2,
                        "unit_price": 589.00,
                    },
                    {
                        "product_name": "Envío",
                        "unit_measure": "UNITS",
                        "quantity": 2,
                        "unit_price": 285.00,
                    },
                ],
            }
        }


class SalesNoteUpdate(BaseModel):
    """Schema for updating a DRAFT sales note."""
    client_id: Optional[int] = None
    client_name: Optional[str] = Field(None, max_length=200)
    payment_method: Optional[str] = Field(None, max_length=30)
    include_taxes: Optional[bool] = None
    notes: Optional[str] = Field(None, max_length=500)


class SalesNoteItemResponse(BaseModel):
    """Line item response."""
    id: int
    product_id: Optional[int] = None
    sku: Optional[str] = None
    product_name: str
    unit_measure: str
    quantity: float
    unit_price: float
    discount_pct: float
    subtotal: float
    ieps_amount: float
    iva_amount: float
    line_total: float

    class Config:
        from_attributes = True


class SalesNoteResponse(BaseModel):
    """Full sales note response."""
    id: int
    note_number: str
    client_id: Optional[int] = None
    client_name: Optional[str] = None
    issuer_name: str
    issuer_rfc: str
    issuer_address: str
    issuer_phone: str
    include_taxes: bool
    subtotal: float
    ieps_total: float
    iva_total: float
    total: float
    total_liters: float
    channel: str
    payment_method: str
    payment_status: str
    status: str
    notes: Optional[str] = None
    created_by: Optional[str] = None
    confirmed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    items: list[SalesNoteItemResponse] = []

    class Config:
        from_attributes = True

File: desert_brew_app/lib/features/sales/domain/entities/sales_note.dart
import 'package:equatable/equatable.dart';

/// Sales note line item.
class SalesNoteItem extends Equatable {
  const SalesNoteItem({
    required this.id,
    required this.productName,
    required this.unitMeasure,
    required this.quantity,
    required this.unitPrice,
    required this.discountPct,
    required this.subtotal,
    required this.iepsAmount,
    required this.ivaAmount,
    required this.lineTotal,
    this.productId,
    this.sku,
  });

  final int id;
  final String productName;
  final String unitMeasure;
  final double quantity;
  final double unitPrice;
  final double discountPct;
  final double subtotal;
  final double iepsAmount;
  final double ivaAmount;
  final double lineTotal;
  final int? productId;
  final String? sku;

  @override
  List<Object?> get props => [id, productName, quantity, unitPrice];
}

/// Sales note entity — the core commercial document.
/// status: DRAFT | CONFIRMED | CANCELLED
/// payment_status: PENDING | PAID | PARTIAL | OVERDUE
class SalesNote extends Equatable {
  const SalesNote({
    required this.id,
    required this.noteNumber,
    required this.issuerName,
    required this.issuerRfc,
    required this.includeTaxes,
    required this.subtotal,
    required this.iepsTotal,
    required this.ivaTotal,
    required this.total,
    required this.totalLiters,
    required this.channel,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.items,
    this.clientId,
    this.clientName,
    this.notes,
    this.createdBy,
    this.confirmedAt,
  });

  final int id;
  final String noteNumber;
  final String issuerName;
  final String issuerRfc;
  final bool includeTaxes;
  final double subtotal;
  final double iepsTotal;
  final double ivaTotal;
  final double total;
  final double totalLiters;
  final String channel;
  final String paymentMethod;
  final String paymentStatus;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<SalesNoteItem> items;
  final int? clientId;
  final String? clientName;
  final String? notes;
  final String? createdBy;
  final DateTime? confirmedAt;

  bool get isDraft => status == 'DRAFT';
  bool get isConfirmed => status == 'CONFIRMED';
  bool get isPaid => paymentStatus == 'PAID';

  @override
  List<Object?> get props => [id, noteNumber, status, paymentStatus, total];
}
